astropy/cosmology/funcs.py

```diff
@@ -253,3 +253,40 @@ def f(z):
         raise CosmologyError(f"Best guess z={res.x} is very close to the lower z limit {zmin}.\n"
                              "Try re-running with a different zmin.")
     return res.x
+
+
+def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500,
+               method='Brent', bracket=None, verbose=False):
+    """``scalar_z_at_value`` for vectorized inputs. Docstring set below.
+
+    The inputs that cannot be vectorized are 'func', 'method', and 'verbose'.
+
+    """
+    # this manual broadcasting is necessary because np.vectorize does not
+    # preserve units.
+
+    shape = np.shape(fval) or (1,)  # shape, at least 1D. For broadcast kwargs
```

Comment:
For progress with this PR, it may be best to not yet count on `np.vectorize` (or `np.frompyfunc`), but you can do the iteration more easily with `np.nditer`. For examples, look in `time/formats.py`.

Reply:
Thanks! I've been tinkering with this for a bit and mostly got it working. The two outstanding issues are: reduction operations s.t. scalar inputs result in scalar output, and allowing broadcast on `bracket` as an object array (b/c it's ragged)

Reply:
I've tried the examples in https://numpy.org/doc/stable/reference/arrays.nditer.html#arrays-nditer to get both working.
1. Using the 'reduce_ok' flag with the reduction operand be initialized to its starting value and resetting the iterator
2. when bracket is 1 object — e.g. `[0, 2]` I do `bracket = np.array([bracket], ], dtype=object)`  but this has shape `(1,2)`, when I want it to have shape `(1,)`.

Reply:
On bracket, since it does contain numbers, maybe easiest to store as a structured array? E.g., 
```
bracket = [0, 2]
np.array(bracket, dtype=float).view([('start', 'f8'), ('stop', 'f8')]).squeeze(-1)
# array((0., 2.), dtype=[('start', '<f8'), ('stop', '<f8')])
bracket = [[0, 2], [3, 4]]
np.array(bracket, dtype=float).view([('start', 'f8'), ('stop', 'f8')]).squeeze(-1)
# array([(0., 2.), (3., 4.)], dtype=[('start', '<f8'), ('stop', '<f8')])
```

Reply:
p.s. Obviously, inside the loop, if you need to get back a list, you can do
```
np.array(bracket, dtype=float).view([('start', 'f8'), ('stop', 'f8')]).squeeze(-1)
bracket[0].item()
(0.0, 2.0)
```

Reply:
unfortunately, the bracket can be length 2 or 3. I think only an object array can support that.

Reply:
Ah, yes, that is why you said ragged array...  Definitely needs an object array. Which means the problem is that `np.array` does not know at what level to place your object... I think this means you are stuck writing your own recursive list comprehension routine.

But how about keeping it simple for now and simply punt this to the user? I.e., allow a list with 2 or 3 elements to indicate a scalar bracket, or a properly formatted object array?

Reply:
Sounds good.

Reply:
Done! both have workable solutions.